### ISSUE TITLE
Allow learning_rate to be a function of time

### ISSUE THREAD

**Issue by studywolf** [nengo/nengo#427]
It would be nice to be able to pass in a function to the learning_rate parameter, so I can have explicit control of it as a function of time or some other variable in my network. e.g.

```
adapt = nengo.Ensemble(100, 1)
system = nengo.Ensemble(100, 1)
answer = nengo.Node(output=training_func)

error = nengo.Ensemble(1, 1)
nengo.Connection(system, error, transform=-1)
nengo.Connection(answer, error)

nengo.Connection(adapt, system, 
                               learning_rule=nengo.PES(error, learning_rate=training_func))
```

where I could define `training_func` as something like

```
def training_func(t):
    return np.exp(-t)
```

or some piecewise thing denoting training epochs or whatever.

As I'm typing this out it occurs to me you could also probably do something equivalent by just scaling the error inside an error node, especially if you want to make it a function of some variable in the network. So maybe that doesn't make sense, but for having learning rates as a function of time this way would probably be more intuitive to program.

Thoughts?

**Comment by tbekolay** [comment]
We discussed this at the lab meeting and the consensus was that this is a good idea ;) And yes, is already possible in `nengo.Voja`, so it would be good to make it possible for all learning rules in a clear way.

We were all agreed that connecting to the learning rule will not work for all rules, as PES already allows you to connect to the learning rule, and it's also not clear when you connect to the learning rule that you're controlling the learning rate. So we want a more obvious solution, likely one that involve modifying `connection.learning_rule.learning_rate` or `connection.learning_rate`.

There wasn't an obvious way to do this, so we're going to use this issue to brainstorm ideas. I'll post a few that come to mind; please add your own, or add thoughts on the ideas I've put forth!

### Option 1: `LearningRuleType.learning_rate` accepts a number or a function of time

In this option, we modify `LearningRuleType` so that the `learning_rate` parameter can accept either a number (as is done now) or a function of time. The builder would have to be modified to run that function during the simulation.

Pros:

- Minimal changes to the frontend.
- Relatively easy to implement.

Cons:

- Cannot accept dynamic input from the simulation (i.e., from a node in the simulation).
- Difficult for some backends to support.

### Option 2: add a `learning_rate` object

This option is a bit less clearly formed in my mind... but the idea is that the `LearningRule` (not the `LearningRuleType`) gains a new attribute, `learning_rate`, that is an instance of an object that can be connected to, similar to `LearningRule` and `Neurons`. In fact, they're similar enough in intention that perhaps all three should inherit from a common base class, or be generalized to some class that works for all three cases.

This is a bigger idea, so the exact pros and cons will depend on the precise proposal. Generally, however, I suspect the pros will be that the API looks nice, and the cons would be that it will require lots of backend changes, and possibly will require a major version bump.

### Option 3: add a `learning_rate` node to the network

In this option, the idea is that a `learning_rate` node is created for every connection (or maybe only connections for which a learning rule has been applied?) This node is somehow special, and is stored as `connection.learning_rate` (or `connection.learning_rule.learning_rate`). The output of the node becomes the learning rate of the learning rule.

This means that, by default, the node should be created such that initially `output=lambda t: learning_rule_type.learning_rate`. If the user wants to make it a function of time, they can modify the `output` parameter of the node. If they want to connect to it, they modify the node's `size_in` and connect to it.

Pros:

- Doesn't require many API changes

Cons:

- Node appears from nowhere magically
- Adds to simulation time when user doesn't want a dynamic learning rate

### Option 4: add a `leanning_rate` parameter to `LearningRule`

This would be new property on the `LearningRule`. By default, it will return `self.learning_rule_type.learning_rate`. It can be set to a value (which will override `self.learning_rule_type.learning_rate`), or a function (which takes in just time input), or a Nengo object (I think just node or ensemble, but I guess probe could work too?) I think it should be pretty clear what happens in each of the three cases, which is maybe a sign that it's a good API.

Pros:

- Clear?
- Doesn't require new object types to be added to the API

Cons:

- Requires relatively large changes to the builder / simulator
- Likely hard for other backends to support

**Comment by jgosmann** [comment]
I argued that we maybe should consider adding named inputs to learning rules in #553 ([code example](https://github.com/nengo/nengo/pull/553#issuecomment-245002724)) which seems to be somewhat related to option 2 here.

**Comment by tcstewar** [comment]
I think I'm not quite sure what the difference is between Option 2 and Option 3, so I'm going to try writing out example code for all the options, to see what it looks like:

###Option 1:

```python
def decay(t):
    return np.exp(-t)
c = nengo.Connection(a, b, learning_rule=nengo.PES(), learning_rate=decay)
nengo.Connection(error, c.learning_rule)
```

###Option 2:

```python
c = nengo.Connection(a, b, learning_rule=nengo.PES())
decay = nengo.Node(lambda t: np.exp(-t))
nengo.Connection(decay, c.learning_rule.learning_rate)
nengo.Connection(error, c.learning_rule)
```

###Option 3:

```python
c = nengo.Connection(a, b, learning_rule=nengo.PES())
decay = nengo.Node(lambda t: np.exp(-t))
nengo.Connection(decay, c.learning_rate)
nengo.Connection(error, c.learning_rule)
```

(Option 3 also specifies that it could instead be ```nengo.Connection(decay, c.learning_rule.learning_rate)```, which is the same as Option 2, I think)

###Option 4:

```python
def decay(t):
    return np.exp(-t)
c = nengo.Connection(a, b, learning_rule=nengo.PES())
c.learning_rule.learning_rate = decay
nengo.Connection(error, c.learning_rule)
```

**Comment by tcstewar** [comment]
If I've got the above options right, then in terms of syntax I really like Option 3.  It's consistent with the general idea of Connections, and I like that it puts the learning rate under the control of the model (and possibly even having neurons control the learning rate....  that makes me feel like it would be consistent with making Connections to ```ens.dopamine```).  Also, I don't think that the "Node appears from nowhere magically", any more than the ```c.learning_rule``` thing appears magically.  (Or at least it's a similar sort of magic).  I also don't think it'll add to the simulation time, since we know at build time whether this exists or not.  So I think implementation-wise, it'll be more like how Option 2 was described -- it's a "thing you can connect to", not an actual Node.

**Comment by tcstewar** [comment]
> we should really allow separate rates for each rule. So for example with Option 3, we'd need a node for each rule (i.e. it should be an attribute on the rule).

Ooo...  yes, that's a very good point.  That would lead me to support something more explicit.  Although what do we do right now for the error signal for multiple learning rules?  Is it ```nengo.Connection(error, conn.learning_rule[0])```?  If so, then I could see ```nengo.Connection(error, conn.learning_rule[0].learning_rate)``` being pretty good.

> I'm not keen on the hidden-node aspect of Option 3. I don't like that the user has to know it's a node in order to modify size_in if they want to connect to it, or overwrite the output function of the node.

Yeah, that seems like something we shouldn't require the user to do.

> To me, the difficult issue is how to allow specifying a constant learning rate easily (i.e. without having to make a constant node and connect it in), and at the same time allow things to connect in.

My hope is that we can do something like "you can set the learning rate to a constant via the constructor or as an attribute (as we currently have it), but if you connection into this special Node-like-thing then it will override that value".  That seems to cover the use cases I can think of, and keeps the cognitive load somewhat minimal....

**Comment by jgosmann** [comment]
I'm at the moment trying to implement learning in the TCM model and one potential approach would benefit from a learning rate that is controlled by another ensemble (but could be implemented without this). That should be possible with options 2 to 4, but I haven't looked in detail which I prefer.

**Comment by jgosmann** [comment]
Somewhat related: I'm now trying a different potential approach that doesn't require a learning rate that changes over time, but has two “error” signals as input. So it would be nice to have two named inputs to the learning rule and it would be nice if the learning rate input would follow the same convention if we do such a thing.

(I think I mentioned somewhere else too that named inputs for learning rules would be nice and that the assumption of a single error signal might not hold always.)
